### PR TITLE
cpu/stm32_common: add DMA implementation for F2/F4/F7

### PR DESCRIPTION
**Pull request by vincent-d**
This is a try to add a generic DMA implementation for f2/f4/f7.

I tested with a stm32f413, UART and SPI drivers seems to work.
If this is accepted, I'll need help to adapt boards configuration and to test.

A `static const dma_conf_t dma_config[]` array is needed in board's `periph_conf.h` to declare which 'physical' streams will be used. Then UART and SPI drivers point to the 'logical' stream (index in `dma_config`).

The DMA API becomes:
```C
/**
 * @brief   Initialize DMA
 */
void dma_init(void);

/**
 * @brief   Execute a DMA transfer
 *
 * This function blocks until the transfer is completed
 *
 * @param[in]  dma     logical DMA stream
 * @param[in]  chan    DMA channel
 * @param[in]  src     source buffer
 * @param[out] dst     destination buffer
 * @param[in]  len     length to transfer
 * @param[in]  mode    DMA mode
 * @param[in]  flags   DMA configuration
 *
 * @return < 0 on error, the number of transfered bytes otherwise
 */
int dma_transfer(dma_t dma, int chan, const void *src, void *dst, size_t len,
                 dma_mode_t mode, uint8_t flags);

/**
 * @brief   Acquire a DMA stream
 *
 * @param[in] dma     logical DMA stream
 */
void dma_acquire(dma_t dma);

/**
 * @brief   Release a DMA stream
 *
 * @param[in] dma     logical DMA stream
 */
void dma_release(dma_t dma);

/**
 * @brief   Start a DMA transfer on a stream
 *
 * @param[in] dma     logical DMA stream
 */
void dma_start(dma_t dma);

/**
 * @brief   Suspend a DMA transfer on a stream
 *
 * @param[in] dma     logical DMA stream
 *
 * @return the remaining number of bytes to transfer
 */
uint16_t dma_suspend(dma_t dma);

/**
 * @brief   Resume a suspended DMA transfer on a stream
 *
 * @param[in] dma     logical DMA stream
 * @param[in] todo    the remaining number of bytes to tranfer
 */
void dma_resume(dma_t dma, uint16_t todo);

/**
 * @brief   Stop a DMA transfer on a stream
 *
 * @param[in] dma     logical DMA stream
 */
void dma_stop(dma_t dma);

/**
 * @brief   Wait for the end of a transfer
 *
 * @param[in] dma     logical DMA stream
 */
void dma_wait_for_end(dma_t dma);

/**
 * @brief   Configure a DMA stream for a new transfer
 *
 * @param[in]  dma     logical DMA stream
 * @param[in]  chan    DMA channel
 * @param[in]  src     source buffer
 * @param[out] dst     destination buffer
 * @param[in]  len     length to transfer
 * @param[in]  mode    DMA mode
 * @param[in]  flags   DMA configuration
 *
 * @return < 0 on error, 0 on success
 */
int dma_configure(dma_t dma, int chan, const void *src, void *dst, size_t len,
                  dma_mode_t mode, uint8_t flags);
```
where `dma_transfer` is a helper function using the other ones when a single transfer is needed.

I wonder whether we could make that API generic for all cpu or not.

### CODE REVIEW

**Comment by smlng** [comment]
Nice! However, considering past issues with changes to SPI, etc. we should test this carefully. 

@aabadie you have lots of nucleos available, right? And maybe @MichelRottleuthner could also help testing, i.e. SPI on nucleos might be interesting to you 😉 I've a nucleo-f401 (or 411?), will report.

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 1]
```diff
@@ -329,6 +381,238 @@ void gpio_init_af(gpio_t pin, gpio_af_t af);
...
+ *
+ * @return the remaining number of bytes to transfer
+ */
+uint16_t dma_suspend(dma_t dma);
+
+/**
+ * @brief   Resume a suspended DMA transfer on a stream
+ *
+ * @param[in] dma     logical DMA stream
+ * @param[in] todo    the remaining number of bytes to tranfer
```
Comment (aabadie) [inline cpu/stm32_common/include/periph_cpu_common.h]: s/tranfer/transfer/

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 2]
```diff
@@ -206,6 +206,51 @@ typedef enum {
 #endif /* ndef DOXYGEN */
 #endif /* ndef CPU_FAM_STM32F1 */
 
+#if defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) || defined(CPU_FAM_STM32F7)
+
+/**
+ * @brief   DMA configuration
+ */
+typedef struct {
+    int stream;            /**< DMA stream */
```
Comment (aabadie) [inline cpu/stm32_common/include/periph_cpu_common.h]: What about having 2 fields: one for stream and one for channel. This would simplify the board configuration I think and only one field has to be added to UART/SPI devices. What do you think ?

File: cpu/stm32_common/periph/dma.c [hunk 3]
```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (C) 2017 OTA keys S.A.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_cortexm_common
```
Comment (aabadie) [inline cpu/stm32_common/periph/dma.c]: should be cpu_stm32_common (which means the same mistake is done in some other periph drivers ;))

File: cpu/stm32_common/periph/dma.c [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+};
+
+static struct dma_ctx dma_ctx[DMA_NUMOF];
+
+static inline uint32_t dma_all_flags(dma_t dma)
+{
+    assert(dma < DMA_NUMOF);
+
+    switch (dma_config[dma].stream & 0x3) {
+    case 0: /* 0 and 4 */
```
Comment (aabadie) [inline cpu/stm32_common/periph/dma.c]: the case block shoud be indented a well (even I don't really like it, that's part of RIOT code conventions...)

File: cpu/stm32_common/periph/dma.c [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    int stream_n = dma_config[dma].stream;
+    uint32_t inc_periph;
+    uint32_t inc_mem;
+
+    DMA_Stream_TypeDef *stream = dma_stream(stream_n);
+    dma_poweron(stream_n);
+    dma_clear_all_flags(dma);
+
+    switch (mode) {
+    case DMA_MEM_TO_MEM:
```
Comment (aabadie) [inline cpu/stm32_common/periph/dma.c]: indent

File: boards/nucleo144-f413/include/periph_conf.h [hunk 6]
```diff
@@ -130,10 +157,8 @@ static const uart_conf_t uart_config[] = {
         .tx_af      = GPIO_AF7,
         .bus        = APB1,
         .irqn       = USART2_IRQn,
-#ifdef UART_USE_DMA
-        .dma_stream = 4,
-        .dma_chan   = 4
-#endif
+        .dma        = 3,
+        .dma_chan   = 4,
```
Comment (aabadie) [inline boards/nucleo144-f413/include/periph_conf.h]: stupid question: how do you select the value used for the channel ? is it global to all streams ?

File: cpu/stm32_common/include/periph_cpu_common.h (hunk 2 above)
Comment (vincent-d) [reply cpu/stm32_common/include/periph_cpu_common.h]: On F2/F4/F7, there is 2 DMAs, each with 8 streams. Each stream have channels. A stream can be used by only one channel at a time. And each peripheral is associated to a given stream/channel (see reference manual for stream/channel). That's why streams are declared in the config, that's the entity which needs to be locked. Declaring the stream/channel couple would make the DMA code a lot more complex because we need a lock per stream (and an interrupt per stream).

File: boards/nucleo144-f413/include/periph_conf.h (hunk 6 above)
Comment (vincent-d) [reply boards/nucleo144-f413/include/periph_conf.h]: Possible stream/channel for each peripheral are described in the reference manual.

File: cpu/stm32_common/periph/dma.c (hunk 4 above)
Comment (vincent-d) [reply cpu/stm32_common/periph/dma.c]: I don't see were in the coding convention we should indent the case. Is there a consensus on that?

Comment (aabadie) [reply cpu/stm32_common/periph/dma.c]: Well looking at other places in the code, the indented case is the most common.

File: cpu/stm32_common/periph/dma.c [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+int dma_configure(dma_t dma, int chan, const void *src, void *dst, size_t len, dma_mode_t mode, uint8_t flags)
+{
+    assert(src != NULL);
+    assert(dst != NULL);
+    assert(dma < DMA_NUMOF);
+
+    int stream_n = dma_config[dma].stream;
+    uint32_t inc_periph;
+    uint32_t inc_mem;
```
Comment (lebrush) [inline cpu/stm32_common/periph/dma.c]: ~~The increments can be deduced from the mode, right? M2M, (`PINC | MINC`) and M2P, P2M (`MINC`). No need to have them in flags. So always `MINC` and only M2M `|= PINC`.~~

Now I see the trick in the spi implementation. Forget about it.

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 8]
```diff
@@ -329,6 +381,238 @@ void gpio_init_af(gpio_t pin, gpio_af_t af);
...
+ */
+uint16_t dma_suspend(dma_t dma);
+
+/**
+ * @brief   Resume a suspended DMA transfer on a stream
+ *
+ * @param[in] dma     logical DMA stream
+ * @param[in] todo    the remaining number of bytes to transfer
+ */
+void dma_resume(dma_t dma, uint16_t todo);
```
Comment (lebrush) [inline cpu/stm32_common/include/periph_cpu_common.h]: `todo` sound strange. What about `remaining`, `bytes`, `length`, ...?

File: cpu/stm32_common/periph/dma.c [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+}
+
+static void dma_clear_all_flags(dma_t dma)
+{
+    assert(dma < DMA_NUMOF);
+
+    DMA_TypeDef *stream = dma_base(dma_config[dma].stream);
+
+    /* Clear all flags */
+    if (dma_hl(dma_config[dma].stream) == 0) {
```
Comment (lebrush) [inline cpu/stm32_common/periph/dma.c]: This can be simplified if #7608 gets merged.

```C
stream->IFCR[dma_hl(dma_config[dma].stream)] = dma_all_flags(dma);
```

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 10]
```diff
@@ -206,6 +206,51 @@ typedef enum {
 #endif /* ndef DOXYGEN */
 #endif /* ndef CPU_FAM_STM32F1 */
 
+#if defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) || defined(CPU_FAM_STM32F7)
```
[the remark this answers is not in the document]
Comment (lebrush) [reply cpu/stm32_common/include/periph_cpu_common.h]: FYI: to support differences in the IRQn I ended up adding some DMA information to the `cpu_conf.h` per family basis.

**Comment by lebrush** [comment]
L4 is channel based, sorry.

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 11]
```diff
@@ -329,6 +381,238 @@ void gpio_init_af(gpio_t pin, gpio_af_t af);
...
+ * @param[in] dma     logical DMA stream
+ */
+void dma_release(dma_t dma);
+
+/**
+ * @brief   Start a DMA transfer on a stream
+ *
+ * @param[in] dma     logical DMA stream
+ */
+void dma_start(dma_t dma);
```
[the remark this answers is not in the document]
Comment (vincent-d) [reply cpu/stm32_common/include/periph_cpu_common.h]: Actually, `dma_transfer` is just a shortcut for `dma_configure`, `dma_start` then `dma_wait`. It's handy when only one channel is used at a time (i.e. uart driver).

File: cpu/stm32_common/include/periph_cpu_common.h (hunk 8 above)
Comment (vincent-d) [reply cpu/stm32_common/include/periph_cpu_common.h]: fixed

File: cpu/stm32_common/include/periph_cpu_common.h [hunk 12]
```diff
@@ -329,6 +381,238 @@ void gpio_init_af(gpio_t pin, gpio_af_t af);
...
+ * @param[in] dma     logical DMA stream
+ */
+void dma_stop(dma_t dma);
+
+/**
+ * @brief   Wait for the end of a transfer
+ *
+ * @param[in] dma     logical DMA stream
+ */
+void dma_wait_for_end(dma_t dma);
```
[the remark this answers is not in the document]
Comment (vincent-d) [reply cpu/stm32_common/include/periph_cpu_common.h]: addressed

**Comment by haukepetersen** [comment]
Nice effort! This is an alternate approach to the functions provided for the `stm32f4`, right?

From a first quick scan through the code it seems quite complex and verbose to me, though. Have you done any code size comparisons so far?

**Comment by vincent-d** [comment]
> Nice effort! This is an alternate approach to the functions provided for the stm32f4, right?

AFAIK there was no dma implementation but just a bunch of functions to set/clear flags and interrupts. These functions were used in the old uart drivers before remodeling.
In this PR, I propose a generic and shared way of accessing dma stream/channels.

> Have you done any code size comparisons so far?

Not really, but a quick try with `nucleo144-f413` give the following results:

- `tests/periph_spi`:

  - master:
   ```
    text	   data	    bss	    dec	    hex	
  15136	    500	   3504	  19140	   4ac4
   ```
  - This PR (with 5 DMA streams configured):
   ```
    text	   data	    bss	    dec	    hex	
  16492	    500	   3564	  20556	   504c
   ```

- `tests/periph_uart`:

  - master:
  ```
   text	   data	    bss	    dec	    hex
  12276	    500	   4724	  17500	   445c
  ```
  - This PR:
  ```
   text	   data	    bss	    dec	    hex
  13472	    500	   4784	  18756	   4944	
  ```

**Comment by haukepetersen** [comment]
Thanks. As I suspected, the code size increase is quite significant. I still wonder if we can reduce the overhead by optimizing the code or if this is the lower bound we have to cope with. If this is the case, I would strongly opt for being able to disable DMA optionally to trade code size for efficiency if needed/wanted.

**Comment by vincent-d** [comment]
>  still wonder if we can reduce the overhead by optimizing the code or if this is the lower bound we have to cope with

I guess some optimizations could be possible, but I think it won't be possible to have dma without any overhead

> I would strongly opt for being able to disable DMA optionally

I agree, I would maybe even tend to disable it by default. This should already be the case, if no `DMA_NUMOF` is defined.

**Comment by vincent-d** [comment]
Adapted to use a submodule, and the "normal" implementation can still be used alongside the DMA one if the periph does not use its `.dma` field (i.e. set it to `DMA_STREAM_UNDEF`).

File: boards/nucleo144-f413/include/periph_conf.h [hunk 13]
```diff
@@ -70,6 +70,27 @@ extern "C" {
 /** @} */
 
 /**
+ * @name    DMA streams configuration
+ * @{
+ */
+static const dma_conf_t dma_config[] = {
```
Comment (haukepetersen) [inline boards/nucleo144-f413/include/periph_conf.h]: I don't see why this should be part of the board configuration. AFAIK the mapping of DMA streams to peripherals is static for each CPU, so there should be no need to configure this on a per-board basis.

Comment (vincent-d) [reply boards/nucleo144-f413/include/periph_conf.h]: This is not that easy, because some peripheral can use different streams and two (or more) peripheral can use the same stream.
Then, we need a mutex per used stream and if we do not want to allocate mutexes for stream which are unused, we need a way to know which streams will be used.
The easiest way I could think was to declare this array in the `periph_conf`. But I would be happy to get rid of it!

**Comment by smlng** [comment]
@vincent-d and @lebrush what about the F1, like in #5095 and #5101?

**Comment by lebrush** [comment]
I think it's better to get the F2/F4/F7 working first which were the ones working already (before all driver unification) and when the DMA API is merged then extend it to the others Fs. What do you think @vincent-d ?

**Comment by vincent-d** [comment]
> What do you think @vincent-d ?

Definitely! F2/F4/F7 are similar, others have a channel based DMA, which could share (hopefully) the same API, but needs more work to adapt.

**Comment by vincent-d** [comment]
How can we proceed with this one?

**Comment by vincent-d** [comment]
Rebased and adapted to remodeling of features

**Comment by vincent-d** [comment]
The commit list became a bit long so I squashed, and rebased on master.

**Comment by smlng** [comment]
@aabadie and @haukepetersen your comments were addressed, clear to move on?

**Comment by aabadie** [comment]
> your comments were addressed, clear to move on?

Yes, but I'd like to test a few boards before. Will it be ok for the release if I do it this next monday ?

**Comment by smlng** [comment]
> Will it be ok for the release if I do it this next monday ?

sure, feature freeze is next Wednesday, so Monday is fine!

File: cpu/stm32f2/cpu.c [hunk 14]
```diff
@@ -19,6 +19,7 @@
 
 #include "cpu.h"
 #include "stmclk.h"
+#include "periph_conf.h"
```
Comment (aabadie) [inline cpu/stm32f2/cpu.c]: This include is not required

File: cpu/stm32f4/cpu.c [hunk 15]
```diff
@@ -18,9 +18,9 @@
  * @}
  */
 
-#include <stdint.h>
 #include "cpu.h"
 #include "stmclk.h"
+#include "periph_conf.h"
```
Comment (aabadie) [inline cpu/stm32f4/cpu.c]: This include is not required

File: cpu/stm32f7/Makefile.features [hunk 16]
```diff
@@ -1,2 +1,3 @@
 FEATURES_PROVIDED += periph_hwrng
+
```
Comment (aabadie) [inline cpu/stm32f7/Makefile.features]: unrelated change

File: cpu/stm32f7/cpu.c [hunk 17]
```diff
@@ -21,6 +21,7 @@
 
 #include "cpu.h"
 #include "stmclk.h"
+#include "periph_conf.h"
```
Comment (aabadie) [inline cpu/stm32f7/cpu.c]: This include is not required

**Comment by vincent-d** [comment]
@aabadie, comments addressed

**Comment by vincent-d** [comment]
I included `periph_cpu.h` from `cpu.c` (needed for `dma_init`)

**Comment by vincent-d** [comment]
Fixed last bugs with uart implementation:
- do not disable dma transfer from irq if the transfer comes from another periph
- acquire the stream before setting `USART_CR3_DMAT` -> factor `acquire` and `release` out of `transfer` function

**Comment by vincent-d** [comment]
- rebased
- added a check for len == 0 in `uart_write`

**Comment by vincent-d** [comment]
@aabadie does your ACK hold? 

ping @haukepetersen!

File: cpu/stm32_common/periph/uart.c [hunk 18]
```diff
@@ -117,24 +117,21 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
 {
-    assert(uart < UART_NUMOF);
-
-    for (size_t i = 0; i < len; i++) {
 #if defined(CPU_FAM_STM32F0) || defined(CPU_FAM_STM32L0) \
     || defined(CPU_FAM_STM32F3) || defined(CPU_FAM_STM32L4) \
     || defined(CPU_FAM_STM32F7)
         while (!(dev(uart)->ISR & USART_ISR_TXE)) {}
-        dev(uart)->TDR = data[i];
+        dev(uart)->TDR = byte;
```
Comment (aabadie) [inline cpu/stm32_common/periph/uart.c]: This line looks over-indented. It should be 4 spaces shifted. Same for the lines above and below.

Comment (vincent-d) [reply cpu/stm32_common/periph/uart.c]: yup, fixed. I also inlined this function and `wait_for_tx_complete`

File: cpu/stm32_common/periph/uart.c [hunk 19]
```diff
@@ -144,6 +141,59 @@ void uart_write(uart_t uart, const uint8_t *data, size_t len)
...
+        return;
+    }
+#endif
+    for (size_t i = 0; i < len; i++) {
+        send_byte(uart, data[i]);
+    }
+    /* make sure the function is synchronous by waiting for the transfer to
+     * finish */
+    wait_for_tx_complete(uart);
+
```
Comment (aabadie) [inline cpu/stm32_common/periph/uart.c]: This empty line could be removed.

File: cpu/stm32_common/periph/uart.c [hunk 20]
```diff
@@ -144,6 +141,59 @@ void uart_write(uart_t uart, const uint8_t *data, size_t len)
...
+            }
+            for (unsigned i = 0; i < len; i++) {
+                send_byte(uart, data[i]);
+            }
+            if (todo > 0) {
+                wait_for_tx_complete(uart);
+                dev(uart)->CR3 |= USART_CR3_DMAT;
+                dma_resume(uart_config[uart].dma, todo);
+            }
+
```
Comment (aabadie) [inline cpu/stm32_common/periph/uart.c]: another empty line (sorry)

**Comment by vincent-d** [comment]
ping @haukepetersen

**Comment by vincent-d** [comment]
Rebased

**Comment by vincent-d** [comment]
Rebased and squashed

**Comment by aabadie** [comment]
Let's go then!

**Comment by smlng** [comment]
Nice!
